### ISSUE TITLE
Transitivity of providedCompile/providedRuntime

### ISSUE THREAD

**Issue by kelemen** [radimk/gradle-nbm-plugin#19]
Given that the plugin will only list the first level dependencies as dependencies of the produced nbm, shouldn't providedCompile/providedRuntime be non-transitive by default? If so, it would also be possible to change 'OpenIDE-Module-Module-Dependencies' to collect dependencies from 'providedRuntime' recursively (+first level compile dependencies for backward compatibility) in case someone wants transitive dependency management for 'OpenIDE-Module-Module-Dependencies'.

**Comment by kelemen** [comment]
I have tried to play around with this and making this change doesn't seem to be trivial because compile.extendsFrom(providedCompile) seems to ignore the transitive property of providedCompile. Regardless, this would be a nice change in my opinion, if there was a way to cleanly define the wanted behaviour.

**Comment by radimk** [comment]
Well, can you define what you want to achieve? I admit that I don't understand what is your goal here.

**Comment by kelemen** [comment]
My main goal would be to cause a compiliation failure if a module was not explicitly added to the providedCompile path but is being used because this will lead the a class not found error  at runtime as this plugin won't declare such modules in 'OpenIDE-Module-Module-Dependencies'.

**Comment by kelemen** [comment]
As a workaround, I have created this internal plugin: [NbmDependencyVerifierPlugin](https://github.com/kelemen/netbeans-gradle-project/blob/master/buildSrc/src/main/groovy/org/netbeans/gradle/build/NbmDependencyVerifierPlugin.java). This is not completely correct but helps detect problems.

**Comment by radimk** [comment]
I don't think it is necessary to have all transitive dependencies enumerated. Also there is a requires/provides style of dependency that can give you access to some classes.

I'm going to close this and we can reopen when we have clearer picture how it should work. This is seconded by the fact that you have a workaround to achieve your goal.

**Comment by kelemen** [comment]
I have tried a couple of more options but it seems that there are other issues making a fix for this infeasible. I have tried to make *compile* not inherit *providedCompile* (the same with *runtime*) and disable transitivity. However, that will mess with annotation processors and Gradle does not make it easy to separate the processor path from the classpath (in fact I had trouble doing so even with adding "-processorpath").

That said, I think it would be better to not let *compile* inherit *providedCompile*. However, that is another issue which I will report if I will encounter any problems caused by this.
